### PR TITLE
New Modal box + Move all settings next to search box to modal box

### PR DESCRIPTION
**Pull request by olamothe**
* Expose lots of Form Component that we were using internally and make them available for external code.

* Disable query syntax by default in the search box. Add a new preference option to allow the end user to set query to on/off

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { ComponentOptions } from '../Base/ComponentOptions';
 import { IComponentBindings } from '../Base/ComponentBindings';
 import { LocalStorageUtils } from '../../utils/LocalStorageUtils';
 import { KEYBOARD } from '../../utils/KeyboardUtils';
-import { PreferencesPanelCheckboxInput, PreferencesPanelTextAreaInput, PreferencePanelMultiSelectInput, IPreferencePanelInputToBuild } from '../PreferencesPanel/PreferencesPanelItem';
+// import { PreferencesPanelCheckboxInput, PreferencesPanelTextAreaInput, PreferencePanelMultiSelectInput, IPreferencePanelInputToBuild } from '../PreferencesPanel/PreferencesPanelItem';
```
Comment (dms1lva) [inline src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts]: Commented imports

File: src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts [hunk 2]
```diff
@@ -119,15 +131,16 @@ export class ResultsFiltersPreferences extends Component {
...
   private preferencePanel: HTMLElement;
   private preferenceContainer: HTMLElement;
-  private preferencePanelCheckboxInput: PreferencesPanelCheckboxInput;
+  private preferencePanelCheckboxInput: { [caption: string]: Checkbox } = {};
   private advancedFilters: HTMLElement;
   private advancedFiltersBuilder: HTMLElement;
-  private advancedFiltersTextInputCaption: PreferencesPanelTextAreaInput;
-  private advancedFiltersTextInputExpression: PreferencesPanelTextAreaInput;
-  private advancedFiltersTabSelect: PreferencePanelMultiSelectInput;
+  private advancedFiltersTextInputCaption: TextInput; // PreferencesPanelTextAreaInput;
```
Comment (dms1lva) [inline src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts]: Comments

File: src/ui/Quickview/Quickview.ts [hunk 3]
```diff
@@ -135,13 +135,12 @@ export class Quickview extends Component {
      * Specifies the Quick View modal window size (width and height) relative to the full window.<br/>
      * The default value is 95% on a desktop and 100% on a mobile device.
      */
-    size: ComponentOptions.buildStringOption({ defaultValue: DeviceUtils.isMobileDevice() ? '100%' : '95%' })
+    size: ComponentOptions.buildStringOption({ defaultValue: DeviceUtils.isMobileDevice() ? '90%' : '90%' })
```
Comment (francoislg) [inline src/ui/Quickview/Quickview.ts]: 😐 

And the doc should be updated.

File: src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts (hunk 1 above)
Comment (francoislg) [inline src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts]: Completely remove those

File: src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts (hunk 2 above)
Comment (francoislg) [inline src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts]: Remove comments

File: src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts [hunk 4]
```diff
@@ -600,6 +614,34 @@ export class ResultsFiltersPreferences extends Component {
       customFilterType: type
     });
   }
+
+  private enlargeModalBox() {
+    const modalBoxContainer = $$(document.body).find('.coveo-modal-container');
+    if (modalBoxContainer) {
```
Comment (francoislg) [inline src/ui/ResultsFiltersPreferences/ResultsFiltersPreferences.ts]: Multiple silent failures here

File: src/ui/ResultsPreferences/ResultsPreferences.ts [hunk 5]
```diff
@@ -110,70 +126,112 @@ export class ResultsPreferences extends Component {
...
 
-  private buildTitle() {
-    var title = $$('div', {
-      className: 'coveo-title'
-    }, l('LinkOpeningSettings'));
-
-    this.element.appendChild(title.el);
+  private buildRadiosInput() {
+    if (this.options.enableQuerySyntax) {
+      const radios = _.map([l('On'), l('Off'), l('Automatic')], (label) => {
```
Comment (francoislg) [inline src/ui/ResultsPreferences/ResultsPreferences.ts]: Would be easier to follow if it was split:

```
const translatedLabels = _.map(['On', 'Off', 'Automatic'], (label) => l(label));
const radios = _.map(translatedLabels, (label) => {
  const radio = this.createRadioButton(label);
  this.preferencePanelRadioInputs[label] = radio
  return radio;
});
```

File: src/ui/ResultsPreferences/ResultsPreferences.ts [hunk 6]
```diff
@@ -110,70 +126,112 @@ export class ResultsPreferences extends Component {
...
-      this.preferencePanelCheckboxInput.select(l('AlwaysOpenInNewWindow'));
+      this.preferencePanelCheckboxInputs[l('AlwaysOpenInNewWindow')].select();
     }
   }
 
-  private fromPreferenceChangeEventToUsageAnalyticsLog(e: Event) {
-    var type = (<HTMLInputElement>e.target).checked ? 'selected' : 'unselected';
-    var preference = (<HTMLInputElement>e.target).value;
+  private fromPreferencesToRadioInput() {
+    if (this.preferences.enableQuerySyntax != null) {
```
Comment (francoislg) [inline src/ui/ResultsPreferences/ResultsPreferences.ts]: You only need 

```
if (this.preferences.enableQuerySyntax === true) {
  this.preferencePanelRadioInputs[l('On')].select();
} else if (this.preferences.enableQuerySyntax === false) {
  this.preferencePanelRadioInputs[l('Off')].select();
} else {
  this.preferencePanelRadioInputs[l('Automatic')].select();
}
```

File: src/ui/ResultsPreferences/ResultsPreferences.ts [hunk 7]
```diff
@@ -110,70 +126,112 @@ export class ResultsPreferences extends Component {
...
     };
-    if (_.contains(selected, l('OpenInOutlookWhenPossible'))) {
-      this.preferences.openInOutlook = true;
-    }
-    if (_.contains(selected, l('AlwaysOpenInNewWindow'))) {
-      this.preferences.alwaysOpenInNewWindow = true;
-    }
+
+    _.each(this.preferencePanelCheckboxInputs, (checkbox: Checkbox, label: string) => {
+      if (this.isSelected(l('OpenInOutlookWhenPossible'), label, checkbox)) {
```
Comment (francoislg) [inline src/ui/ResultsPreferences/ResultsPreferences.ts]: `OpenInOutlookWhenPossible`, `AlwaysOpenInNewWindow`, `On`, `Off`, `Automatic` come back often, they should be in constants.

File: src/ui/ResultsPreferences/ResultsPreferences.ts [hunk 8]
```diff
@@ -110,70 +126,112 @@ export class ResultsPreferences extends Component {
...
+    });
+    _.each(this.preferencePanelRadioInputs, (radio: RadioButton, label: string) => {
+      if (this.isSelected(l('On'), label, radio)) {
+        this.preferences.enableQuerySyntax = true;
+      }
+      if (this.isSelected(l('Off'), label, radio)) {
+        this.preferences.enableQuerySyntax = false;
+      }
+      if (this.isSelected(l('Automatic'), label, radio)) {
+        delete this.preferences.enableQuerySyntax;
```
Comment (francoislg) [inline src/ui/ResultsPreferences/ResultsPreferences.ts]: Here you `delete` but later you set to `null`. Pick one :)

File: src/ui/ShareQuery/ShareQuery.ts [hunk 9]
```diff
@@ -147,21 +147,37 @@ export class ShareQuery extends Component {
...
   }
 
+  private buildCompleteQuery() {
+    this.completeQuery = <HTMLInputElement>$$('input', {
+      type: 'text',
+      className: 'coveo-share-query-summary-info-input'
+    }).el;
+
+    const lastQuery = this.queryController.getLastQuery();
+    this.completeQuery.value = Utils.trim(this.outputIfNotNull(lastQuery.q) + ' ' + this.outputIfNotNull(lastQuery.aq) + ' ' + this.outputIfNotNull(lastQuery.cq));
```
Comment (francoislg) [inline src/ui/ShareQuery/ShareQuery.ts]: `Utils.trim(_.filter([lastQuery.q, lastQuery.aq, lastQuery.cq], (queryPart) => this.outputIfNotNull(queryPart)).join(' '));` ?

File: src/utils/Dom.ts [hunk 10]
```diff
@@ -131,7 +131,9 @@ export class Dom {
    * Empty the element and all childs from the dom;
    */
   public remove(): void {
-    this.el.parentNode.removeChild(this.el);
+    if (this.el.parentNode) {
```
Comment (francoislg) [inline src/utils/Dom.ts]: Does nothing when `parentNode` is undefined?

File: test/ui/TextInputTest.ts [hunk 11]
```diff
@@ -40,6 +44,31 @@ export function TextInputTest() {
         textInput.setValue(value);
         expect(textInput.getValue()).toEqual(value);
       });
+
+      it('should call the onchange', () => {
+        textInput.setValue('test');
+        expect(spy).toHaveBeenCalledWith(textInput);
+      });
+
+      it('should not call the onchange multiple time', () => {
```
Comment (francoislg) [inline test/ui/TextInputTest.ts]: Missing tests for 

* should call onChange twice with two `setValue`
* should call onChange with `setValue` and empty value

?

File: src/GlobalExports.ts [hunk 12]
```diff
@@ -27,3 +27,12 @@ export function lazyExport(component: IComponentDefinition, promiseResolve: Func
   Initialization.registerAutoCreateComponent(component);
   promiseResolve(component);
 }
+
+export function lazyExportModule(mod: any, promiseResolve: Function) {
+  if (mod.doExport) {
+    mod.doExport();
+  } else {
+    new Logger(this).error(`Module ${mod} has no export function !`);
```
Comment (olamothe) [inline src/GlobalExports.ts]: I'll remove that, it should not force a type.

File: src/GlobalExports.ts [hunk 13]
```diff
@@ -27,3 +27,12 @@ export function lazyExport(component: IComponentDefinition, promiseResolve: Func
   Initialization.registerAutoCreateComponent(component);
   promiseResolve(component);
 }
+
+export function lazyExportModule(mod: any, promiseResolve: Function) {
```
Comment (olamothe) [inline src/GlobalExports.ts]: Actually, I should remove the error.

It should not throw, nor force a specific type.

File: src/ui/Base/QueryBuilder.ts [hunk 14]
```diff
@@ -102,11 +102,11 @@ export class QueryBuilder {
    */
   public enableQuestionMarks: boolean;
   /**
-   * Whether to disable the special query syntax such as field references for the basic query expression (parameter q).
+   * Whether to enable the special query syntax such as field references for the basic query expression (parameter q).
    * It is equivalent to a No syntax block applied to the basic query expression.
    * If not specified, the parameter defaults to false.
    */
-  public disableQuerySyntax: boolean = false;
+  public enableQuerySyntax: boolean = false;
```
Comment (olamothe) [inline src/ui/Base/QueryBuilder.ts]: Yes.

File: src/GlobalExports.ts (hunk 13 above)
Comment (olamothe) [inline src/GlobalExports.ts]: I hate the pattern of returning an empty promise.

The pattern of returning a promise is useful if you actually return something that can be modified down the line.

File: src/utils/Dom.ts (hunk 10 above)
Comment (olamothe) [inline src/utils/Dom.ts]: Yes. It's already removed.

File: src/ui/Quickview/Quickview.ts (hunk 3 above)
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: I'll just remove that useless option.

**Pull request merged**